OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/fieldaccess/reflection/ReflectionRelatedFieldAccessesAnalysis.scala

```diff
@@ -0,0 +1,1055 @@
+                )
+
+                val matchers = data._1.matchers +
+                    retrieveSuitableMatcher[Set[ObjectType]](
+                        classOpt.map(Set(_)),
+                        data._1.pc,
+                        v => new ClassBasedFieldMatcher(v, !data._1.matchers.contains(PublicFieldMatcher))
+                    )
+
+                addFieldAccess(state.callContext, data._1.pc, _ => data._1.receiver, _ => data._1.param, matchers)
+            }
+
+        if (eps.isFinal) state.removeDependee(eps.toEPK)
+        else state.updateDependency(eps)
+
+        returnResult(fieldVar)
+    }
+
+    protected def handleGetField(
+        context:         ContextType,
+        accessPC:        Int,
+        fieldDefSite:    Int,
+        actualReceiver:  AccessReceiver,
+        actualParameter: Option[AccessParameter],
+        baseMatchers:    Set[FieldMatcher],
+        stmts:           Array[Stmt[V]]
+    )(implicit indirectFieldAccesses: IndirectFieldAccesses, state: ReflectionState[ContextType]): Set[FieldMatcher] = {
+        def constructClassBasedFieldMatcherForClassVar(
+            classVar:                 V,
+            currentMatchers:          Set[FieldMatcher],
+            onlyFieldsExactlyInClass: Boolean
+        ): FieldMatcher = {
+            MatcherUtil.retrieveClassBasedFieldMatcher(
+                context,
+                classVar,
+                ClassDepender(accessPC, actualReceiver, actualParameter, currentMatchers, classVar, stmts),
+                accessPC,
+                stmts,
+                project,
+                () => failure(accessPC, actualReceiver, actualParameter, currentMatchers),
+                onlyFieldsExactlyInClass,
+            )
+        }
+
+        var matchers = baseMatchers
+        stmts(fieldDefSite).asAssignment.expr match {
+            case call @ VirtualFunctionCall(_, ObjectType.Class, _, "getDeclaredField" | "getField", _, receiver, params) =>
+                val isGetField = call.name == "getField"
+                if (isGetField) matchers += PublicFieldMatcher
+
+                if (!matchers.contains(NoFieldsMatcher))
+                    matchers += MatcherUtil.retrieveNameBasedFieldMatcher(
+                        context,
+                        params.head.asVar,
+                        NameDepender(accessPC, actualReceiver, actualParameter, matchers, params.head.asVar, stmts, receiver.asVar, context),
+                        accessPC,
+                        stmts,
+                        () => failure(accessPC, actualReceiver, actualParameter, matchers)
+                    )
+
+                if (!matchers.contains(NoFieldsMatcher))
+                    matchers += constructClassBasedFieldMatcherForClassVar(receiver.asVar, matchers, !isGetField)
+
+            case call @ VirtualFunctionCall(_, ObjectType.Class, _, "getDeclaredFields" | "getFields", _, receiver, _) =>
```

Comment:
I'm still unsure whether it is sensible to handle this here. This can never happen on the initial call of `handleGetField`, only for the recursion in the ArrayLoad case, right?

Reply:
Well if also foreach loops over arrays are compiled to array loads (which would make sense) then yes. To access a field via any reflective getter or setter, you need to load it first, which when using `getDeclaredFields` or similar always requires choosing an element from the returned array.

Reply:
As discussed in person, we will move handling of such fields to a seperate methods that is called when an array load is encountered.